### PR TITLE
Dynamic type webhooks

### PR DESCRIPTION
**Pull request by slok**
Fixes https://github.com/slok/kubewebhook/issues/36

This PR adds the ability to use dynamic webhooks. This webhooks infer the type of the request based on the received review JSON data. If the type can't be inferred because is unknown then it will fallback to `runtime.Unstructured` object type.

We consider unknown types (core objects that are not tracked by the webhook kubernetes libraries and CRDs).

These dynamic webhooks will be activated when the webhook factory receives a configuration without the `Obj` field set.

The webhooks have been refactored internally to use a internal `ObjectCreator` that is the one responsible to create the object from the received raw JSON, so we have 2, a `Static` and `Dynamic` one.

These webhooks should be also very helpful on multitype webhooks or webhooks that mutate/validate metadata fields like labels.

### CODE REVIEW

**Comment by turkenh** [comment]
Looking forward for this PR. 
Is there an ETA?

**Comment by slok** [comment]
> Looking forward for this PR.
> Is there an ETA?

Hey @turkenh Sorry fo not merging it in 3 days, I normally merge once I make the PR and passes the CI, in this case I wanted to test it a little bit more, so I was setting up a production webhook with so I could test it for a while in a production environment, I've had the webhook ~32h hours mutating multiple types and seems to be ok, in mutation and in resource consumption, so I'll merge into master.

If you start using it and detect problems or bugs, please open an issue so I can check and fix it.

Thanks!

**Comment by turkenh** [comment]
@slok thanks a lot!

Testing with a validating webhook for `DELETE` requests and getting the following error:

```
Error from server (InternalError): Internal error occurred: failed calling webhook "delete-validating-webhooks": an error on the server ("{\"response\":{\"uid\":\"3c1de137-8371-4ba8-a6d1-3bd6aeae511f\",\"allowed\":false,\"status\":{\"metadata\":{},\"status\":\"Failure\",\"message\":\"unexpected end of JSON input\"}}}") has prevented the request from succeeding
```

Any ideas?

**Comment by turkenh** [comment]
Could this be related to not having request body for delete requests? 
Related: https://github.com/slok/kubewebhook/issues/41

File: pkg/webhook/validating/webhook.go [hunk 1]
```diff
@@ -83,33 +83,30 @@ func NewWebhook(cfg WebhookConfig, validator Validator, ot opentracing.Tracer, r
...
-func (w *staticWebhook) Review(ctx context.Context, ar *admissionv1beta1.AdmissionReview) *admissionv1beta1.AdmissionResponse {
+func (w validateWebhook) Review(ctx context.Context, ar *admissionv1beta1.AdmissionReview) *admissionv1beta1.AdmissionResponse {
 	w.logger.Debugf("reviewing request %s, named: %s/%s", ar.Request.UID, ar.Request.Namespace, ar.Request.Name)
 
-	obj := helpers.NewK8sObj(w.objType)
-	runtimeObj, ok := obj.(runtime.Object)
-	if !ok {
-		err := fmt.Errorf("could not type assert metav1.Object to runtime.Object")
+	// Create a new object from the raw type.
+	runtimeObj, err := w.objectCreator.NewObject(ar.Request.Object.Raw)
```
Comment (turkenh) [inline pkg/webhook/validating/webhook.go:97]: Should we use `ar.Request.oldObject.Raw` is it is a delete operation?

**Comment by turkenh** [comment]
Just tested and same webhook works for `UPDATE` requests.

**Comment by slok** [comment]
Yes, is related with #41.

I opened an issue https://github.com/slok/kubewebhook/issues/63

I'll try fixing this in the following days! many thanks for the feedback!!

**Comment by slok** [comment]
@turkenh I just merged the fix in #64.

Could you please, test the `DELETE` operation on your validation webhook?

Let's see if now it doesn't fail on unmarshaling, you can access the object in the webhook, validate the deletion object, and see if it works correctly all the process.

Thanks!

**Comment by turkenh** [comment]
@slok yes, I can verify the issue is fixed on my side and I can now use dynamic type webhook with `DELETE` operation.

Thanks a lot for quick fix!

**Comment by slok** [comment]
Awesome!

I'll publish a release this week with these changes.

Many thanks for your help and feedback!

**Pull request merged**